### DISCUSSION TITLE
테라폼관련 질문이 있습니다!

### DISCUSSION THREAD

**Discussion by jellychoco** [tedilabs/fastcampus-devops#60] [Q&A] [answered]
안녕하세요! 현재 쿠버네티스 관련 영상을 보고있습니다.

Part1. Docker 기초 ch0.1 파트개요 및 실습준비
-08. 테라폼코드를 이용하여 미리 준비된 AWS실습환경 구성 편을 보고 있습니다.

terraform 에 대한 설명이 없었던거 같지만, 검색해서 init, apply 까지는 했습니다만
그 다음엔 어찌해야할지 영상에 나오지 않은것 같아 문의드립니다.
apply 해서 aws에 vpc, instance 생성되는것 보여주시는데, 
보는사람은 어떻게 설정해야 aws 에 vpc, instance 가 생성되는지는 안나오는것 같습니다.

<img width="964" alt="스크린샷 2022-04-06 오후 10 42 43" src="https://user-images.githubusercontent.com/52277404/161988573-ad4d3b2b-7fad-416f-ba95-51d2fb32b77c.png">

기존에 다른분이 했던 질문에 보면 영상이 있다고 답변이 되어있는데 제가 보는 강의에는 없는거 같습니다. 
아니면 terraform 은 강사님만 사용하시는거고, 저희는 직접 vpc 생성해서 인스턴스 만들어야 하나요~?

**Comment by posquit0** [comment, marked as answer]
안녕하세요~
우선 불편을 드려서 죄송합니다.

실제 제가 진행한 강의는 [초격차 패키지 : 한 번에 끝내는 AWS 인프라 구축과 DevOps 운영](https://fastcampus.co.kr/dev_online_awsdevops) 인데, 패스트캠퍼스의 또 다른 강의인 [Kubernetes 와 Docker로 한 번에 끝내는 컨테이너 기반 MSA](https://fastcampus.co.kr/dev_online_kubemsa) 강의의 첫 도커 기초 파트로 활용이 되면서 해당 이슈가 발생한 것 같습니다.

해당 Discussion 스레드의 다른 글들은 기존 강의에 대한 질문응답이다보니 혼란을 더 야기한 듯 합니다.

우선 문제 해결에 도움을 드리기 위해 답변을 작성해보니 참고 부탁드립니다.

<img width="1409" alt="image" src="https://user-images.githubusercontent.com/1484002/162003267-ac64bdc9-e721-4238-9a4e-7fa7c9c78bdb.png">

- 강의 자료의 7페이지를 보시면 실습 환경을 세 가지 중 하나를 택일하도록 안내하고 있습니다.
- 설명하는 바와 같이 컨테이너를 다루는 환경 구성은 컨테이너 기술이 리눅스 기술이다보니 리눅스(우분투 등) 환경을 권장하지만, 맥이나 윈도우 WSL에서 docker를 설치하셔서 진행하셔도 강의 진행에 크게 문제가 되지는 않습니다.
  - 볼륨 이나 네트워크 관련 내용을 다룰 때 이슈를 겪을 수 있습니다.
- 수강자 분께서 리눅스 환경에 docker, docker-compose를 설치할 수 있다면 어떠한 환경에서 진행을 하든 문제가 되지는 않습니다.
  - 말씀주신 것처럼 테라폼 없이 직접 EC2 인스턴스를 설치 후 진행하시는 것도 가능합니다. 혹시나 테라폼을 이용해보고 싶으시다면 아래 내용을 참고 부탁드립니다! :)
- 해당 강의에서 제공하는 테라폼 코드는 손 쉽게 AWS 계정 내에 docker / docker-compose가 설치된 Ubuntu 리눅스 서버를 EC2 이미지로 구동하기 위해 제공해주는 코드입니다. 기존 초격차 강의에서는 테라폼을 함께 다루다보니 해당 강의에 섞여 들어간 듯 합니다... 😢 
  - https://github.com/tedilabs/fastcampus-devops/tree/main/3-docker-kubernetes/env/terraform-aws-ubuntu 디렉토리 내에 `ec2-instance` 디렉토리와 `network` 디렉토리가 존재하는 것을 확인할 수 있습니다.
  - `network` 디렉토리 내에서 우선 `terraform init`으로 테라폼 워크스페이스 초기화 후 `terraform apply`를 진행하여 네트워크(VPC) 구성을 진행합니다.
  - `ec2-instance` 디렉토리 내에서 `terraform init`으로 테라폼 워크스페이스 초기화 후 `terraform apply`를 진행하면 이전 단계에서 생성한 네트워크 위에 EC2 인스턴스를 실행하여 도커 컨테이너 실습 환경을 구성합니다.
  - 테라폼을 실행하기에 앞서 수강자 분의 AWS 계정에 대한 인증 정보 설정을 진행해야 합니다.
    - [Terraform AWS Provider: Authentication and Configuration](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#authentication-and-configuration) 문서를 참조하실 수 있습니다.
    - 기본적으로 AWS 계정 내에서 발급한 Access Key ID와 Secret Access Key 쌍을 환경변수로 설정하여 진행하실 수 있습니다.
    - 첨부해주신 오류 화면이 AWS 계정에 대한 인증정보가 설정되어 있지 않아 발생하는 오류입니다.

혹시나 더 도움이 필요하시다면 커멘트 남겨주시기 바랍니다! :)
